docs/proposals/read-write-operations-bucket.md

```diff
@@ -0,0 +1,244 @@
+
+To match partial upload safeguards we want to delete block in reverse order:
+
+> 7 . To schedule delete operation, delete `meta.json` file. All components will exclude this block and compactor will do eventual deletion assuming the block is partially uploaded. [Compactor change needed]
+ 
+We schedule deletions instead of doing them straight away for 3 reasons: 
+* Readers that have loaded this block can still access index and metrics for some time. On next sync they will notice lack of meta.json and assume partial block which excludes it from block being loaded.
+* Only compactor deletes metrics and index.
+* In further delete steps, starting with meta.json first ensures integrity mark being deleted first, so in case of deletion process being stopped, we can treat this block as partial block (rule 4th) and delete it gracefully.
+
+There might be exception for malformed blocks that blocks compaction or reader operations. Since we may need to unblock the system 
+immediately the block can be forcibly removed meaning that query failures may occur (reader loaded block, but not aware block was deleted).
+
+> 8 . Compactor waits minimum 15m (`deleteDelay`) before deleting the whole `To Delete` block. [Compactor change needed]
+
+This is to make sure we don't forcibly remove block which is still loaded on reader side. We do that by counting time from
+spotting lack of meta.json first. After 15 minutes we are ok to delete the whole directory.
+
+## Diagram
+
+Below diagram shows block states defined by our rules.
+
+[![Diagram](https://docs.google.com/drawings/d/e/2PACX-1vSZjddrWBkf2ynczeIMSzGo-yqe7Cg9X-TG_LjyWAP_ToyZolMEG_UnbNlvoSetRhDLmlfABd_rJ4W9/pub?w=960&h=720)](https://docs.google.com/drawings/d/1leRN6GUvU98Ur2TowMefk4ycTFNBRi9t1-dGtOPmHg0/edit?usp=sharing)
+
+## Risks
+
+* What if someone will set too low `syncDelay`? And upload itself will take more than `syncDelay` time. Block will be assumed as `ToDelete` state and will be removed. Thanks of `deleteDelay` it might have still some time to recover.How to ensure that will not happen? 
+  * Minimum syncDelay? 
+  * Grab time for syncDelay from object meta instead? Not from ULID? That is quite hard.
+  * Add safeguard on upload path? `upload timeout` has to smaller than `syncDelay`.
+
+* What if one block is malformed. Readers cannot handle it and crashes. How repair procedure will work? We can have repair process that can download block locally, rewrite it and fix it and upload. Problem is that it will take `syncDelay` time to
+appear in system. Since we are blocked, we need to make the block available immediately, ignoring eventual consistency limitations.
+Potential solutions:
+  * Just delete problematic block and live with 15 minute of delay for not having some portion of metrics? Will it not cause any downsampling issues?
```

Comment:
If this would cause alerts to go off that should not, I would refrain from just pulling the data...

Reply:
What do you mean? (: